### ISSUE TITLE
Scale model using target

### ISSUE THREAD

**Issue by gmiguelr87** [openMVG/openMVG#1320]
Dear all,

I'm trying to scale automatically the resulting pointclouds using a target. For that, I detect the corners of the target and I insert their x,y coordinates in the .feat files and its respective match in the matching.f.txt file in order to measure its distance in the 3D coordinates and be able to scale because I know the real distance between corners, obvious thing.
When the sparse.ply is generated, my "problem" is that I dont know how to calculate the position of the 3D coordinates of the corners in the .ply file.
Could you help me to understand how is the sparse.ply generated depending on the feat/matching data?

Thank you in advance!

**Comment by pmoulon** [comment]
You can use the Ground Control Point (GCP) functionality to do that.

Basically, the idea is that you defined some 3d point with know 3d coordinates and known 2d image projection.
Then OpenMVG allows you to compute the Scale, Rotation and Translation to move the scene to your desired coordinates system.

See https://github.com/openMVG/openMVG/blob/master/docs/sphinx/rst/software/ui/SfM/control_points_registration/GCP.rst

You can also see here a demo about using ARUco market to do a similar thing:
https://github.com/openMVG/openMVG/pull/1239

Note: Computing the 3D location of the points according to the image observations and camera location is called triangulation. You can look [here](https://github.com/openMVG/openMVG/blob/master/src/openMVG/sfm/sfm_data_triangulation.hpp) to see the two sfm_data triangulation engine.

**Comment by gmiguelr87** [comment]
Thank you @pmoulon !
I'm trying to use your reference to #1239 
I understand that I have to apply that in the sfm_data.json generated with openMVG_main_SfMInit_ImageListing and then insert this new file into openMVG_main_IncrementalSfM, right?

Because I'm not using Aruco market, I'm trying to generate the json file, but when I call to the executable Control_Points_Registration, it returns me values inf and -nan(ind).

Please, could someone check my file to explain me if it's ok?

[sfm_data.zip](https://github.com/openMVG/openMVG/files/2054260/sfm_data.zip)

The values of the corners of the targets are (order: upper_right, upper_left, down_left, down_right):
# Line format: img_name tag_id x1 y1 x2 y2 x3 y3 x4 y4
"DSC_7640.JPG" 0 3036.03 2204.29 2495.9 2310.18 2638.59 2841.59 3214.72 2710.49
"DSC_7641.JPG" 0 2627.57 2314.03 2031.81 2285.92 1969.48 2862.24 2615.06 2890.26
"DSC_7642.JPG" 0 2346.2 2322.66 1607.68 2196.28 1271.95 2849.68 2101.88 3025.11
"DSC_7643.JPG" 0 3340.16 2143.44 2507.51 2334.7 2876.31 3108 3822.11 2817.65

**Comment by gmiguelr87** [comment]
I answer to myself...

After seeing more information about this, it is used after openMVG_main_IncrementalSfM.exe, it is this sfm_data file that needs to be modified. I measure it, but when I insert it in the Control_Points_Registration executable I get the message "Insufficient number of triangulated control points".
I know it's a message from an external executable, but if anyone can help me I would be very grateful.

Thank you!

**Comment by pmoulon** [comment]
You need to set image observations for the control point in at least two view. If you can set the GCP observations in 3 images or more it would be even better.

Triangulation needs at least 2 image observations.

**Comment by pmoulon** [comment]
I forgot to mention that you need to run the binary Control_Points_Registration on the sfm_data.bin results you obtained with openMVG_main_IncrementalSfM. You need valid camera poses to perform triangulations.

**Comment by gmiguelr87** [comment]
Hi @pmoulon !!

Thank you again! I finally got the scaled point cloud.
Regarding your explanation, I know that at least 2 observations are needed, my mistake was that I intended to proccess the sfm_data.json before openMVG_main_IncrementalSfM and also by a failure in the json syntax.

Attached is the json file in case it's helpful to anyone.
This file includes as ground control points the four corners of a target detected in the four images being processed.
[sfm_data.zip](https://github.com/openMVG/openMVG/files/2058669/sfm_data.zip)
